src/internals/iterator.rs

```diff
@@ -0,0 +1,126 @@
+use std::ptr;
+use yara_sys::{YR_MEMORY_BLOCK, YR_MEMORY_BLOCK_ITERATOR};
+
+#[derive(Debug)]
+pub struct MemoryBlock<'a> {
+    /// base contains the base address of the current block
+    base: u64,
+    /// size contains the size of the current block
+    size: u64,
+    /// data is used to read size bytes into a byte slice
+    data: &'a [u8],
+}
+
+impl<'a> MemoryBlock<'a> {
+    pub fn new(base: u64, size: u64, data: &'a [u8]) -> Self {
+        Self { base, size, data }
+    }
+
+    fn as_yara(&mut self) -> YR_MEMORY_BLOCK {
+        let fetch_data = if self.size == 0 {
+            mem_block_fetch_data_null
+        } else {
+            mem_block_fetch_data
+        };
+
+        YR_MEMORY_BLOCK {
+            base: self.base,
+            size: self.size,
+            context: self as *mut MemoryBlock as *mut std::os::raw::c_void,
+            fetch_data: Some(fetch_data),
+        }
+    }
+}
+
+pub trait MemoryBlockIterator {
+    fn first(&mut self) -> Option<MemoryBlock>;
+    fn next(&mut self) -> Option<MemoryBlock>;
+}
+
+pub trait MemoryBlockIteratorSized: MemoryBlockIterator {
+    fn file_size(&mut self) -> u64;
+}
+
+#[derive(Debug)]
+pub struct WrapperMemoryBlockIterator<T> {
+    iter: T,
+    mem_block: std::mem::MaybeUninit<YR_MEMORY_BLOCK>,
+}
+
+impl<T> WrapperMemoryBlockIterator<T> {
+    pub fn new(iter: T) -> Self {
+        Self {
+            iter,
+            mem_block: std::mem::MaybeUninit::uninit(),
```

Comment:
I am no sure with this. I needed to store current memory_block. After this PR minimal supported rustc version is 1.55

Reply:
Why? MaybeUninit is stable since 1.36. Am I missing something?

Reply:
I use MaybeUninit.write, which was stabilized at 1.55